Lib/pdb.py

```diff
@@ -2491,6 +2499,490 @@ def set_trace(*, header=None, commands=None):
+                        # Interact mode has been cancelled client-side,
+                        # likely by a ^D EOF at the prompt.
+                        self._interact_state = None
+                        return line
+                    case {"interact": str(lines)}:
+                        if self._interact_state:
+                            return lines
+                        # Otherwise, fall through to report an error and loop.
+                    case {
+                        "completion": {
+                            "text": str(text),
+                            "line": str(line),
+                            "begidx": int(begidx),
+                            "endidx": int(endidx),
+                        }
+                    }:
+                        items = self._complete_any(text, line, begidx, endidx)
+                        self._send(completion=items)
+                        continue
+            except json.JSONDecodeError:
+                pass
+            # Invalid JSON, or doesn't meet the schema, or wrong PDB state.
+            self.error(f"Ignoring invalid remote command: {payload}")
+
+    def _complete_any(self, text, line, begidx, endidx):
+        if begidx == 0:
+            return self.completenames(text, line, begidx, endidx)
+
+        cmd = self.parseline(line)[0]
+        if cmd:
+            compfunc = getattr(self, "complete_" + cmd, self.completedefault)
+        else:
+            compfunc = self.completedefault
+        return compfunc(text, line, begidx, endidx)
+
+    def cmdloop(self, intro=None):
+        self.preloop()
+        if intro is not None:
+            self.intro = intro
+        if self.intro:
+            self.message(str(self.intro))
+        stop = None
+        while not stop:
+            if self.cmdqueue:
+                line = self.cmdqueue.pop(0)
+            else:
+                if not self._command_name_cache:
+                    self._command_name_cache = self.completenames("", "", 0, 0)
+                    self._send(command_list=self._command_name_cache)
+
+                mode = "interact" if self._interact_state else "pdb"
+                self._send(prompt=self.prompt, mode=mode)
+                line = self._read_command()
+                if self._interact_state is not None:
+                    self._run_in_python_repl(line)
+                    continue
+            line = self.precmd(line)
+            stop = self.onecmd(line)
+            stop = self.postcmd(stop, line)
+        self.postloop()
+
+    def detach(self):
+        # Detach the debugger and close the socket without raising BdbQuit
+        self.quitting = False
+        if self._owns_sockfile:
+            self._sockfile.close()
+
+    def do_EOF(self, arg):
+        ret = super().do_EOF(arg)
+        self.detach()
+        return ret
+
+    def do_q(self, arg):
+        ret = super().do_q(arg)
+        self.detach()
+        return ret
+
+    def do_quit(self, arg):
```

Comment:
If you control the `cmdloop` already, you don't need to overwrite these commands. Just check `self.quitting` when you exit the while loop - if that's `True`, do `detach`.

Reply:
Ah, even simpler, we can use the `postloop` hook. Thanks, that's much cleaner! 557a7259bebe80660c84f36fd27a14543fe2eb85